src/Magnum/Platform/GlfwApplication.cpp

```diff
@@ -0,0 +1,214 @@
+
+    /* Set context version, if requested */
+    if(configuration.version() != Version::None) {
+        Int major, minor;
+        std::tie(major, minor) = version(configuration.version());
+        glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, major);
+        glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, minor);
+        #ifndef MAGNUM_TARGET_GLES
+        if(configuration.version() >= Version::GL310) {
+            glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GLFW_TRUE);
+            glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
+        }
+        #else
+        glfwWindowHint(GLFW_CLIENT_API, GLFW_OPENGL_ES_API);
+        #endif
+    }
+
+    /* Set context flags */
+    _window = glfwCreateWindow(configuration.size().x(), configuration.size().y(), configuration.title().c_str(), monitor, nullptr);
+    if(!_window) {
+        Error() << "Platform::GlfwApplication::tryCreateContext(): cannot create context";
+        glfwTerminate();
+        return false;
+    }
+
+    /* Set callbacks */
+    glfwSetFramebufferSizeCallback(_window, staticViewportEvent);
+    glfwSetKeyCallback(_window, staticKeyEvent);
+    glfwSetCursorPosCallback(_window, staticMouseMoveEvent);
+    glfwSetMouseButtonCallback(_window, staticMouseEvent);
+    glfwSetScrollCallback(_window, staticMouseScrollEvent);
+
+    glfwMakeContextCurrent(_window);
+    glfwSwapInterval(1);
+
+    /* Return true if the initialization succeeds */
+    return _context->tryCreate();
+}
+
+GlfwApplication::~GlfwApplication() {
+    glfwDestroyWindow(_window);
+    glfwTerminate();
+}
+
+int GlfwApplication::exec() {
+    while(!glfwWindowShouldClose(_window)) {
+        if(_needsRedraw) {
+            drawEvent();
+        }
+        glfwPollEvents();
+    }
+    return 0;
+}
+
+void GlfwApplication::staticKeyEvent(GLFWwindow*, int key, int, int action, int) {
+    KeyEvent e(static_cast<KeyEvent::Key>(key));
+
+    if(action == GLFW_PRESS) {
+        _instance->keyPressEvent(e);
+    } else if(action == GLFW_RELEASE) {
+        _instance->keyReleaseEvent(e);
+    } /* we don't handle GLFW_REPEAT */
+}
+
+void GlfwApplication::staticMouseMoveEvent(GLFWwindow*, double x, double y) {
+    MouseMoveEvent e{Vector2i{Int(x), Int(y)}};
+    _instance->mouseMoveEvent(e);
+}
+
+void GlfwApplication::staticMouseEvent(GLFWwindow*, int button, int action, int) {
+    MouseEvent e(static_cast<MouseEvent::Button>(button));
+
+    if(action == GLFW_PRESS) {
+        _instance->mousePressEvent(e);
+    } else if(action == GLFW_RELEASE) {
+        _instance->mouseReleaseEvent(e);
+    } /* we don't handle GLFW_REPEAT */
+}
+
+void GlfwApplication::staticMouseScrollEvent(GLFWwindow*, double xoffset, double yoffset) {
+    MouseScrollEvent e(Vector2d{xoffset, yoffset});
+    _instance->mouseScrollEvent(e);
+
+    if(yoffset != 0.0) {
+        MouseEvent e1((yoffset > 0.0) ? MouseEvent::Button::WheelUp : MouseEvent::Button::WheelDown);
+        _instance->mousePressEvent(e);
```

Comment:
This line seems somehow superfluous, there should be just one press event, right? Or am I missing something?